src/main/java/com/google/api/codegen/discogapic/DiscoGapicProvider.java

```diff
@@ -28,30 +29,55 @@
   private final Document document;
   private final GapicProductConfig productConfig;
   private final CommonSnippetSetRunner snippetSetRunner;
-  private final DocumentToViewTransformer transformer;
+  private final DocumentToViewTransformer documentTransformer;
+  private final SchemaToViewTransformer schemaTransfomer;
 
   private DiscoGapicProvider(
       Document document,
       GapicProductConfig productConfig,
       CommonSnippetSetRunner snippetSetRunner,
-      DocumentToViewTransformer transformer) {
+      DocumentToViewTransformer documentTransformer,
+      SchemaToViewTransformer schemaTransfomer) {
     this.document = document;
     this.productConfig = productConfig;
     this.snippetSetRunner = snippetSetRunner;
-    this.transformer = transformer;
+    this.documentTransformer = documentTransformer;
+    this.schemaTransfomer = schemaTransfomer;
   }
 
   public List<String> getSnippetFileNames() {
-    return transformer.getTemplateFileNames();
+    return documentTransformer.getTemplateFileNames();
   }
 
+
   public Map<String, Doc> generate() {
-    return generate(null);
+    Map<String, Doc> results = new TreeMap<>();
+    results.putAll(generate(null));
+    results.putAll(generateSchemas(null));
+    return results;
   }
 
   public Map<String, Doc> generate(String snippetFileName) {
-    List<ViewModel> surfaceDocs = transformer.transform(document, productConfig);
+    List<ViewModel> surfaceDocs = documentTransformer.transform(document, productConfig);
+
+    Map<String, Doc> docs = new TreeMap<>();
+    for (ViewModel surfaceDoc : surfaceDocs) {
+      if (snippetFileName != null && !surfaceDoc.templateFileName().equals(snippetFileName)) {
+        continue;
+      }
+      Doc doc = snippetSetRunner.generate(surfaceDoc);
+      if (doc == null) {
+        // generation failed; failures are captured in the model.
+        continue;
+      }
+      docs.put(surfaceDoc.outputPath(), doc);
+    }
+
+    return docs;
+  }
 
+  public Map<String, Doc> generateSchemas(String snippetFileName) {
```

Comment:
Don't duplicate the code between this new function (`generateSchemas`) and the existing one (`generate`).